syntax = "proto3";

package raft;

service Raft {
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
    rpc ClientRequest(ClientRequestMessage) returns (ClientResponseMessage);
}

message LogEntry {
    int32 term = 1;
    string command = 2;  // Command to be executed by the state machine
}

message AppendEntriesRequest {
    int32 term = 1;
    int32 leaderId = 2;
    int32 prevLogIndex = 3;
    int32 prevLogTerm = 4;
    repeated LogEntry entries = 5;  // Empty for heartbeat
    int32 leaderCommit = 6;
}

message AppendEntriesResponse {
    int32 term = 1;
    bool success = 2;
}

message RequestVoteRequest {
    int32 term = 1;
    int32 candidateId = 2;
    int32 lastLogIndex = 3;
    int32 lastLogTerm = 4;
}

message RequestVoteResponse {
    int32 term = 1;
    bool voteGranted = 2;
}

message ClientRequestMessage {
    string command = 1;
}

message ClientResponseMessage {
    bool success = 1;
    string message = 2;
    string leaderId = 3;
}
